### PR TITLE
support multi pd server

### PR DESCRIPTION
**Pull request by bobotu**
PTAL @coocood

### CODE REVIEW

File: pd/client.go [hunk 1]
```diff
@@ -77,111 +86,136 @@ type client struct {
...
-		go c.reportRegionHeartbeat(ctx, stream, errCh, wg)
-		go c.receiveRegionHeartbeat(ctx, stream, errCh, wg)
+		go c.reportRegionHeartbeat(errCh, wg)
+		go c.receiveRegionHeartbeat(errCh, wg)
 		select {
 		case err := <-errCh:
 			log.Warnf("[%s][pd] heartbeat stream get error: %s ", c.tag, err)
-			cancel()
-			time.Sleep(time.Second)
+			for {
```
Comment (coocood) [inline pd/client.go]: The loop is too deeply nested, how about extract a method?

File: pd/client.go [hunk 2]
```diff
@@ -77,111 +86,136 @@ type client struct {
...
-	cc, err := grpc.Dial(strings.TrimPrefix(c.url, "http://"), grpc.WithInsecure())
-	if err != nil {
-		return nil, errors.Trace(err)
+func (c *client) doRequest(ctx context.Context, f func(context.Context, pdpb.PDClient) error) error {
+	var err error
+	var retry int
+	for {
+		c.connMu.RLock()
+		ctx, cancel := context.WithTimeout(ctx, pdTimeout)
+		err = f(ctx, pdpb.NewPDClient(c.clientConn))
```
Comment (coocood) [inline pd/client.go]: Can we move the f call out of the lock?

File: pd/client.go [hunk 3]
```diff
@@ -61,13 +62,21 @@ var (
...
-	url        string
-	tag        string
-	clusterID  uint64
-	clientConn *grpc.ClientConn
+	url       []string
+	tag       string
+	clusterID uint64
+
+	connMu        sync.RWMutex
+	clientConn    *grpc.ClientConn
```
Comment (coocood) [inline pd/client.go]: Can we maintain all the connections for each URL like the way https://github.com/pingcap/pd/blob/master/client/client.go do?

File: pd/client.go [hunk 4]
```diff
@@ -77,111 +86,136 @@ type client struct {
...
+		c.connMu.RLock()
+		ctx, cancel := context.WithTimeout(ctx, pdTimeout)
+		err = f(ctx, pdpb.NewPDClient(c.clientConn))
+		c.connMu.RUnlock()
+		cancel()
+		if err == nil {
+			return nil
+		}
+		
+		for {
```
Comment (cfzjywxk) [inline pd/client.go]: maybe wrap the retry logic into a `ReConnectWithRetry` function?

File: pd/util.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		if clusterID != 0 {
+			if cid != uint64(clusterID) {
+				return nil, nil, errors.Errorf("PD response clusterID mismatch, want %d, got %d", clusterID, cid)
+			}
+		} else {
+			clusterID = cid
+		}
+
+		if members == nil {
+			members = resp
```
Comment (cfzjywxk) [inline pd/util.go]: could we break the loop here, if we have got the members successfully

**Comment by bobotu** [comment]
PTAL @coocood @cfzjywxk

**Comment by coocood** [comment]
LGTM

**Comment by cfzjywxk** [comment]
LGTM better to add some comments for new functions

**Comment by coocood** [comment]
@cfzjywxk we can do it later

**Pull request merged**